dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,256 @@
+-- let baz = "qux"
+-- ----------------
+-- Input for this function is:
+-- [(FP 0 4, "foo"), (FP 1 4, "baz")]
+-- And it will be transformed into:
+-- [(FP 0 4, Tag "let foo " "foo"), (FP 1 20, Tag "let baz " "baz")]
+-- where 20 is byte offset from file start.
+fixPosAndDefinition :: Text -> [(PosInFile, Text)] -> [(PosInFile, Tag)]
+fixPosAndDefinition t = foldMap (\(FP ln c, term) ->
+             let (ln', offset, tPattern) = fromMaybe (fallbackInfoForText ln c)
+                                                     (infoForText term ln)
+             in [(FP ln' offset, Tag tPattern term)])
+    where mls :: M.Map Int (Int, Text) 
+          -- ^ mls is map that for each line has length of file before this map and line content.
+          --   In example above, first line is 15 bytes long and '\n', mls contain:
+          --   (1, (16, "let foo = "bar"")
+          --   That allow us to get byte offset easier.
+          mls = M.fromList . fst .
+                foldl (\(ls, lsl) (n, l) ->
+                           let lsl' = lsl + 1 + lengthInBytes l in
+                           ((n, (lsl, l)):ls, lsl')) ([], 0) . zip [1..] $ T.lines t
+          lineFromMap ln = fromMaybe (0, "") (ln `M.lookup` mls)
+          lengthInBytes = BS.length . encodeUtf8
+          {-| get information about term from map of lines
+              In most cases, PosInFile after getTagsFromExpr point to byte before term.
+              It's better to have term in term pattern, so this function finds and updates
+              PosInFile and generate pattern.
+          -} 
+          infoForText
+              :: Text
+              -- ^ term to find
+              -> Int
+              -- ^ line where to start
+              -> Maybe (Int, Int, Text)
+              -- ^ (Line number, byte offset, pattern to find term in file)
+          infoForText _ 0 = Nothing
+          infoForText term ln
+              | T.null part2 = infoForText term (ln - 1)
+              | otherwise = Just (ln, lsl + 1 + lengthInBytes part1, part1 <> termAndNext)
+              where (lsl, l) = lineFromMap ln
+                    (part1, part2) = T.breakOn term l
+                    termAndNext = T.take (T.length term + 1) part2
+          fallbackInfoForText ln c = (ln, lsl + 1 + lengthInBytes pat, pat)
+              where (lsl, l) = lineFromMap ln
+                    pat = T.take c l
+
+getTagsFromExpr :: Expr Src a -> [(PosInFile, Text)]
+getTagsFromExpr = go (FP 0 0) []
+    where go lpos mts = \case
+              (Let b e) -> go lpos (mts <> parseBinding lpos b) e
+              (Annot e1 e2) -> go lpos (go lpos mts e1) e2
+              (Record mr) -> mts <> tagsFromDhallMap lpos mr
+              (RecordLit mr) -> mts <> tagsFromDhallMap lpos mr
+              (Union mmr) -> mts <> tagsFromDhallMapMaybe lpos mmr
+              (Note s e) -> go (srcToPosInFile s) mts e
+              _ -> mts
+          tagsFromDhallMap lpos = foldMapWithKey (tagsFromDhallMapElement lpos)
```

Comment:
Can you add empty lines between the separate definitions?

Reply:
Added